onboarding_project/squad_pantry_app/models.py

```diff
@@ -150,3 +149,62 @@ class ConfigurationSettings(models.Model):
 
     def __str__(self):
         return self.constant
+
+
+class PerformanceMetrics(models.Model):
+    created_at = models.DateTimeField(auto_now_add=True, unique=True)
+    average_throughput = models.IntegerField(editable=False)
+    average_turnaround_time = models.CharField(max_length=256, editable=False)
+
+    @classmethod
+    def calculate_avg_performance_metrics(cls):
+        if PerformanceMetrics.objects.all().count() < 1:
+            completed_orders_curr_day = Order.objects.filter(
+                status=Order.DELIVERED,
+                closed_at__isnull=False
+            )
+        else:
+            start_date_time = PerformanceMetrics.objects.latest('id').created_at
+            completed_orders_curr_day = Order.objects.filter(
+                status=Order.DELIVERED,
+                closed_at__isnull=False,
+                created_at__range=(start_date_time, timezone.now())
+            )
+        throughput = len(completed_orders_curr_day)
+
+        if throughput == 0:
+            return 0, '00:00:00.00'
+
+        total_turnaround_time = 0
+        for obj in completed_orders_curr_day:
+            total_turnaround_time += (obj.closed_at - obj.created_at).total_seconds()
+
+        average_turnaround_time = total_turnaround_time/throughput
+        return timedelta(seconds=average_turnaround_time), throughput
+
+    @classmethod
+    def create_avg_performance_metrics(cls):
+        turnaround_time, throughput = PerformanceMetrics.calculate_avg_performance_metrics()
+        print(turnaround_time, throughput)
+        PerformanceMetrics.objects.create(average_throughput=throughput, average_turnaround_time=turnaround_time)
+
+    @classmethod
+    def get_metrics_data(cls, start_date, end_date):
+        metrics = PerformanceMetrics.objects.filter(created_at__range=(start_date, end_date))
+        num_obj = len(metrics)
+        total_throughput = 0
+        total_turnaround_time = 0
+        for obj in metrics:
```

Comment:
You are doing the aggregation in python. Is there a way of doing the same thing on the db? Why would we want to do it on the db as opposed to the current method?

Reply:
We discussed this yesterday. If we do a `count()` on the `queryset` it hits the DB again. So, when we need `count` and `queryset` we should `filter` the `queryset` and do `len()`.